lisa/tests/staging/sched_android.py

```diff
@@ -0,0 +1,204 @@
+
+    @classmethod
+    def _from_target(cls, target, res_dir, ftrace_coll=None):
+        # The boost hold feature for CFS should help keep the frequency stable
+        state = target.sched.get_features()['SCHEDTUNE_BOOST_HOLD_ALL']
+
+        # XXX - Context manager for this ? Move it to RTATestBundle ?
+        target.sched.set_feature('SCHEDTUNE_BOOST_HOLD_ALL', True)
+        obj = super()._from_target(target, res_dir, ftrace_coll=ftrace_coll)
+        target.sched.set_feature('SCHEDTUNE_BOOST_HOLD_ALL', state)
+
+        return obj
+
+    @classmethod
+    def get_rtapp_profile(cls, plat_info):
+        cpu = plat_info['capacity-classes'][-1][0]
+        rtapp_profile = {}
+        rtapp_profile['rta_stune'] = Periodic(
+            duty_cycle_pct = 1, # very small task, no impact on freq w/o boost
+            duration_s = 10,
+            period_ms = 16,
+            cpus = [cpu] # pin to big CPU, to focus on frequency selection
+        )
+
+        return rtapp_profile
+
+    def trace_window(self, trace):
+        rta_start, rta_stop = super().trace_window(trace)
+
+        cpu = self.plat_info['capacity-classes'][-1][0]
+        fdf = trace.df_events('cpu_frequency')
+        fdf = fdf[fdf.cpu == cpu]
+
+        # Find the frequency events before and after the task runs
+        freq_start = fdf[fdf.index < rta_start].index[-1]
+        freq_stop = fdf[fdf.index > rta_stop].index[0]
+
+        return (freq_start, freq_stop)
+
+    @requires_events('cpu_frequency')
+    def test_frequency(self, freq_margin_pct=10) -> ResultBundle:
+        """
+        Test that frequency selection followed the boost
+
+        :param: freq_margin_pct: Allowed margin between estimated and measured
+            average frequencies
+        :type freq_margin_pct: int
+
+        Compute the expected frequency given the boost level and compare to the
+        real average frequency from the trace.
+        Check that the difference between expected and measured frequencies is
+        no larger than ```freq_margin_pct```.
+        """
+        cpu = self.plat_info['capacity-classes'][-1][0]
+        freqs = self.plat_info['freqs'][cpu]
+        max_freq = max(freqs)
+
+        # Estimate the target frequency, including sugov's margin, and round
+        # into a real OPP
+        target_freq = min(max_freq, max_freq * self.boost / 80)
+        target_freq = list(filter(lambda f: f >= target_freq, freqs))[0]
+
+	# Get the real average frequency
+        avg_freq = self.trace.analysis.frequency.get_average_cpu_frequency(cpu)
+
+        distance = abs(target_freq - avg_freq) * 100 / target_freq
+        res = ResultBundle.from_bool(distance < freq_margin_pct)
+        res.add_metric("target freq", target_freq, 'kHz')
+        res.add_metric("average freq", avg_freq, 'kHz')
+
+        return res
+
+class BoostedMaxFreq(SchedTuneFreqBoost):
+    """
+    A tiny task with 100% boost
+    """
+    boost = 100
+
+class BoostedFreq80(SchedTuneFreqBoost):
```

Comment:
Depending on the way we want to analyze the results, it may be advantageous to use the same pattern as `load_tracking.Invariance`. One class is in charge of deciding how many individual items are created, and the tests are done on each item individually. If necessary, there can be an "aggregated" version of the test that checks all items as well.

This avoids the proliferation of classes and ABCs, and more importantly, allows deciding "on the fly" (based on plat_info) how many items are needed and what are their `from_target()` parameters.

Reply:
Yeah it is somewhat similar, all those results need to be looked at as a whole. I'll try and see if I can re-use the `load_tracking.Invariance` pattern.

Reply:
To map Invariance to your new class, you can apply this correspondance:
* `boost` becomes a parameter of your `XXXItem.from_target()` factory, just like `cpu` and `freq` for `InvarianceItem.from_target()`.
* Look at `Invariance.test_task_util_avg` and `Invariance._test_all_freq` for the aggregated version. Ignore `test_cpu_invariance` and `test_freq_invariance` since they do more than just checking an XXXItem test is passing on all items.
* Once we are done, we can add the bit in exekall_customize to tag the XXXItem objects, so they show up with their `boost` value in the non-aggregated results.